File: hub/src/light/effect/color_effects/triple_linear_color_transition.py
from light.effect.partial_effect import PartialEffect
from light.effect.resolved_effect import ResolvedEffect
from util.color_transition import ColorTransition
from light.effect.effect_priority import EffectPriority

class TripleLinearColorTransition(PartialEffect):
    def __init__(self, srcRgb, dstRgb1, dstRgb2, duration):
        super().__init__(0)
        self.src = srcRgb
        self.dst1 = dstRgb1
        self.dst2 = dstRgb2
        self.duration = duration
    
    def getEffect(self, t):
        dt = t - self.startTime
        tNorm = dt / self.duration

        dst = self.dst1 if dt <= 0.5 else self.dst2
        src = self.src if dt <= 0.5 else self.dst1
        tNormAdjusted = (tNorm * 2) if dt <= 0.5 else ((tNorm - 0.5) * 2)

        return ResolvedEffect.createRgbw(ColorTransition.linear(tNormAdjusted, src, dst))

    def isTemporal(self):
        return False

    def isComplete(self, t):
        return False

File: hub/drumlite-js/src/light/effect/color/TripleLinearColorTransition.ts
import PartialEffect, { EffectParameters, EffectParameter } from "../PartialEffect"
import RGB from "../../RGB"
import ResolvedEffect from "../../../effect/ResolvedEffect";
import ColorTransition from "../../../util/ColorTransition";

export class TripleLinearColorTransitionParams extends EffectParameters {
    effectName = "Triple Linear Transition";
    typeName = "Color";

    constructor(src: RGB = new RGB(), dst1: RGB = new RGB(), dst2: RGB = new RGB(), duration: number = 1) {
        super(0);
        this.params.src = new EffectParameter<RGB>("Start Color", src);
        this.params.dst1 = new EffectParameter<RGB>("Intermediate Color", dst1);
        this.params.dst2 = new EffectParameter<RGB>("End Color", dst2);
        this.params.duration = new EffectParameter<number>("Duration", duration);
    }
}

export default class TripleLinearColorTransition extends PartialEffect<TripleLinearColorTransitionParams> {
    constructor(params: TripleLinearColorTransitionParams, dt = 0) {
        super(params, dt);
    }
    
    public getEffect(t: number) {
        const {
            src,
            dst1,
            dst2,
            duration
        } = this.params.params;

        const dt = t - this.params.params.startTime.val;
        const tNorm = dt / (duration.val);

        const dst = (dt <= 0.5 ? dst1.val : dst2.val);
        const srcRgb = (dt <= 0.5 ? src.val : dst1.val);
        const tNormAdjusted = dt <= 0.5 ? (tNorm * 2) : ((tNorm - 0.5) * 2);

        return ResolvedEffect.createRgb(ColorTransition.linear(tNormAdjusted, srcRgb, dst))
    }

    public isTemporal() {
        return false;
    }

    public isComplete(t: number) {
        return false;
    }
}